users/views.py

```diff
@@ -0,0 +1,82 @@
+from rest_framework import status
+from rest_framework.decorators import api_view, permission_classes
+from . models import User
+from rest_framework_simplejwt.tokens import RefreshToken
+from django.contrib.auth.signals import user_logged_in
+from rest_framework.generics import RetrieveUpdateAPIView
+from rest_framework.permissions import IsAuthenticated
+
+def get_tokens_for_user(user):
+    refresh = RefreshToken.for_user(user)
+
+    return {
+        'refresh': str(refresh),
+        'access': str(refresh.access_token),
+    }
+
+
+class CreateUserAPIView(APIView):
+    # Allow any user (authenticated or not) to access this url
+    permission_classes = (p.AllowAny,)
+
+    def post(self, request):
+        user = request.data
+        serializer = UserSerializer(data=user)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+
+@api_view(['POST'])
+@permission_classes([p.AllowAny, ])
+def authenticate_user(request):
+    try:
+        email = request.data['email']
+        password = request.data['password']
+        user = User.objects.get(email=email, password=password)
+        if user:
+            try:
+                tokens = get_tokens_for_user(user)
+                user_details = {'name': "%s %s" % (
+                    user.first_name, user.last_name), 'token': tokens["access"], 'refresh_token': tokens["refresh"]}
+                #TODO figure out purpose of below
+                user_logged_in.send(sender=user.__class__,
+                                    request=request, user=user)
+                return Response(user_details, status=status.HTTP_200_OK)
+            except Exception as e:
+                raise e
+        else:
+            res = {
+                'error': 'can not authenticate with the given credentials or the account has been deactivated'}
+            return Response(res, status=status.HTTP_401_UNAUTHORIZED)
+    except KeyError:
+        res = {'error': 'please provide a email and a password'}
+        return Response(res)
+
+
+class UserRetrieveUpdateAPIView(RetrieveUpdateAPIView):
+    # Allow only authenticated users to access this url
+    permission_classes = (IsAuthenticated,)
+    serializer_class = UserSerializer
+
+    def get(self, request, *args, **kwargs):
+        # serializer to handle turning our `User` object into something that
+        # can be JSONified and sent to the client.
+        serializer = self.serializer_class(request.user)
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def put(self, request, *args, **kwargs):
+        serializer_data = request.data.get('user', {})
+
+        serializer = UserSerializer(
+            request.user, data=serializer_data, partial=True
+        )
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
```

Comment:
don't forget to add an empty line at the end of the file

Reply:
Done.